homeassistant/components/homekit/type_cameras.py

```diff
@@ -0,0 +1,422 @@
+        ]
+
+        video_options = {
+            "codec": {
+                "profiles": [
+                    VIDEO_CODEC_PARAM_PROFILE_ID_TYPES["BASELINE"],
+                    VIDEO_CODEC_PARAM_PROFILE_ID_TYPES["MAIN"],
+                    VIDEO_CODEC_PARAM_PROFILE_ID_TYPES["HIGH"],
+                ],
+                "levels": [
+                    VIDEO_CODEC_PARAM_LEVEL_TYPES["TYPE3_1"],
+                    VIDEO_CODEC_PARAM_LEVEL_TYPES["TYPE3_2"],
+                    VIDEO_CODEC_PARAM_LEVEL_TYPES["TYPE4_0"],
+                ],
+            },
+            "resolutions": resolutions,
+        }
+        audio_options = {
+            "codecs": [
+                {"type": "OPUS", "samplerate": 24},
+                {"type": "AAC-eld", "samplerate": 16},
+            ]
+        }
+
+        self.streaming_status = STREAMING_STATUS["AVAILABLE"]
+        self.has_srtp = True
+        self.stream_address = self.config.get(CONF_STREAM_ADDRESS) or get_local_ip()
+        try:
+            ipaddress.IPv4Address(self.stream_address)
+            self.stream_address_isv6 = b"\x00"
+        except ValueError:
+            self.stream_address_isv6 = b"\x01"
+        self.sessions = {}
+
+        if self.config[CONF_SUPPORT_AUDIO]:
+            self.add_preload_service(SERV_MICROPHONE)
+        management = self.add_preload_service(SERV_CAMERA_RTP_STREAM_MANAGEMENT)
+        management.configure_char(
+            CHAR_STREAMING_STATUS, getter_callback=self._get_streaming_status
+        )
+        management.configure_char(
+            CHAR_SUPPORTED_RTP_CONFIGURATION,
+            value=PyhapCamera.get_supported_rtp_config(True),
+        )
+        management.configure_char(
+            CHAR_SUPPORTED_VIDEO_STREAM_CONFIGURATION,
+            value=PyhapCamera.get_supported_video_stream_config(video_options),
+        )
+        management.configure_char(
+            CHAR_SUPPORTED_AUDIO_STREAM_CONFIGURATION,
+            value=PyhapCamera.get_supported_audio_stream_config(audio_options),
+        )
+        management.configure_char(
+            CHAR_SELECTED_RTP_STREAM_CONFIGURATION,
+            setter_callback=self.set_selected_stream_configuration,
+        )
+        management.configure_char(
+            CHAR_SETUP_ENDPOINTS, setter_callback=self.set_endpoints
+        )
+
+    def update_state(self, new_state):
+        """Handle state change to update HomeKit value."""
+        pass
+
+    def _get_streaming_status(self):
+        return PyhapCamera._get_streaimg_status(self)
+
+    async def _start_stream(self, objs, reconfigure):
+        return await PyhapCamera._start_stream(self, objs, reconfigure)
+
+    async def _stop_stream(self, objs):
+        return await PyhapCamera._stop_stream(self, objs)
+
+    def set_selected_stream_configuration(self, value):
+        """Set the selected stream configuration."""
+        return PyhapCamera.set_selected_stream_configuration(self, value)
+
+    def set_endpoints(self, value):
```

Comment:
This looks largely like the upstream one, but you've implemented one of their TODO's. There is a lot of duplicated code though. Do you think we can get it upstream?